standards/pre-rootfs-environment.rst

```diff
@@ -0,0 +1,201 @@
+
+Pre-rootfs environments shall respect the configuration of the HSD.
+
+Android partition mounts
+------------------------
+
+The pre-rootfs environment shall mount all partitions that are required to run Android on the Device. These partitions shall be mounted into the ``/android`` directory of the rootfs. They generally include block devices which back the filesystem in Android and include mount points such as ``/persist``, ``/firmware``, and ``/efs``. These would become ``/android/persist``, ``/android/firmware``, and ``/android/efs`` in the HSD rootfs.
+
+Generally, each partition on the device is mounted by reading the ``fstab`` file generated by the Android build system. This file should be retrieved from the Halium-provided Android system image at each boot.
+
+Special cases
+^^^^^^^^^^^^^
+
+The ``/system`` and ``/data`` Android filesystems are special cases. They must be mounted differently than the remaining block device mounts.
+
+``/system``
+"""""""""""
+
+``/system`` shall be the filesystem provided by the Android ``system.img``. 
+
+``/data``
+"""""""""
+
+``/data`` shall be a location on non-volatile, writable storage. The location shall not contain the rootfs' user data storage. This prevents the Android container being able to access the Device owner's data.
+
+.. note::
+
+    In practice, this can be achieved by mounting the physical ``/data/android-data`` at the HSD's ``/android/data``. The Device owner's data can then safely reside at the root of ``/data``.
+
+``/userdata``
+-------------
+
+A non-volatile and writable location to store user data shall be mounted at the rootfs' ``/userdata`` mount point.
+
+.. note::
+
+    Most pre-rootfs environments choose to mount the Device's physical ``/data`` partition at the rootfs' ``/userdata``.
+
+Kernel modules mount
+--------------------
+
+The pre-rootfs environment shall mount Linux kernel modules stored at ``/android/system/lib/modules`` into the rootfs. The target of these mounts shall be specified by the rootfs in the file ``/etc/halium/modules``.
+
+The rootfs' configuration file for this behavior may include the text ``$version``, which is a variable that shall be replaced by the running Linux kernel's release identifier. This identifier is generally found by running the command ``uname -r``.
+
+The following is an example of the ``/etc/halium/modules`` file, which specifies that the pre-rootfs environment should mount kernel modules at the rootfs' ``/lib/modules/[VERSION]``.
+
+.. code-block:: text
+   :name: /etc/halium/modules
+   
+    /lib/modules/$version
+
+In this case, if the kernel identifier is ``4.13.11-31.current``, the target of the Linux kernel module mount shall be ``/lib/modules/4.13.11-31.current/`` on the rootfs.
+
```

Comment:
What should pre-rootfs do if the target directory doesn't exist? Should it create the directory if the rootfs is writable? How about when rootfs is not writable? (Target directory needs to exist to be bind-mountable)

Reply:
Hmm, that's a good point. The Halium docs say to make the Android side do any required `modprobe`-ing. Maybe we should continue with that and remove all module mounting form the standard?